### PR TITLE
binary_perimeter

### PR DESCRIPTION
**Pull request by jacobkimmel**
## Description
[Tell us about your new feature, improvement, or fix! If relevant, please supplement the PR with images.]

Added a function binary_perimeter() to find the perimeter of objects in a binary input image in the morphology.binary module.
Takes a binary image and selem neighborhood as inputs. Outputs a binary image of perimeters.
Added a unit test for this function in morphology.binary.test_binary()

This function is useful for many image segmentation tasks.

While I realize it is somewhat simple to script up, I believe having a function in place is beneficial for those migrating from commercial toolboxes, where even simple combinations of morphological operators like this are included as standalone functions. 

**Example output:**

cellmask
cellmask_perim

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

### CODE REVIEW

**Comment by sciunto** [comment]
a little question. What would be the difference with a canny filter?

**Comment by jacobkimmel** [comment]
Hey @sciunto,

Canny filtering using the default skimage.feature implementation does not provide exact perimeters, as it first smooths with a Gaussian filter, then morphologically thins edges and performs hystersis. 

This naturally leads to perimeters that are slightly "inside" the original mask. I would assume they'll be 1/2 the size of the Gaussian filter inside in fact, as the maximum values post-filtering will be 1/2 filter width inside the actual positive binary object.

The morphological padding, erosion, unpadding, subtraction algorithm implemented here finds the *actual* edges of objects in the mask.

This is easiest to show with a couple examples.

`from skimage import feature`
`edges = feature.canny(mask)`  
`perim = binary_perimeter(mask) # as implemented in this PR`
`imgplt = imshowpair(mask, edges)`
`plt.show()`

**Canny Edges Overlaid with Original Mask**

cannyoverlay

*Zoomed In*

cannyoverlay_zoom


`imgplt = imshowpair(mask, perim)`
`plt.show()`

**binary_perimeter() Overlaid with Original Mask**

bwperimoverlay

*Zoomed In*

bwperimoverlay_zoom

File: skimage/morphology/binary.py [hunk 1]
```diff
@@ -144,3 +144,35 @@ def binary_closing(image, selem=None, out=None):
     dilated = binary_dilation(image, selem)
     out = binary_erosion(dilated, selem, out=out)
     return out
+
+@default_selem
+def binary_perimeter(image, selem=None):
+    """Return perimeter mask of objects in a binary image
```
Comment (sciunto) [inline skimage/morphology/binary.py]: full stop at the end of the sentence

File: skimage/morphology/binary.py [hunk 2]
```diff
@@ -144,3 +144,35 @@ def binary_closing(image, selem=None, out=None):
...
+
+    Find the perimeters of objects in a provided binary image for a given
+    connectivity
```
Comment (sciunto) [inline skimage/morphology/binary.py]: full stop. Remove "provided".

File: skimage/morphology/binary.py [hunk 3]
```diff
@@ -144,3 +144,35 @@ def binary_closing(image, selem=None, out=None):
...
+    Uses padding and binary erosion
```
Comment (sciunto) [inline skimage/morphology/binary.py]: Put this information in a Notes section

File: skimage/morphology/binary.py [hunk 4]
```diff
@@ -144,3 +144,35 @@ def binary_closing(image, selem=None, out=None):
...
+
+    Returns
+    -------
+    perim : binary image of perimeters of objects in bw
+
+    """
+
+    bw_pad = np.pad(image, pad_width=1, mode='constant', constant_values=0)
+    bw_eroded = binary_erosion(bw_pad, selem=selem)
+    bw_eroded = bw_eroded[1:-1,1:-1] # unpad output
```
Comment (sciunto) [inline skimage/morphology/binary.py]: this line can be concatenated with the one above

**Comment by soupault** [comment]
Isn't `measure.regionprops` a right place to add this into?
I'm not saying that the proposed implementation is of no use, but I think that our current approach (use `measure.label` to label regions, `measure.regionprops` to get properties of them) suits the needs better.

**Comment by sciunto** [comment]
Naturally, I wouldn't go in regionprops if I would seek for such features. Also, regionprops is crowded.

I see this more as a filter than a property of a region.

**Comment by jni** [comment]
@jacobkimmel I believe the function "segmentation.mark_boundaries" already does this?

File: skimage/morphology/tests/test_binary.py [hunk 5]
```diff
@@ -45,6 +45,14 @@ def test_binary_opening():
     grey_res = img_as_bool(grey.opening(bw_img, strel))
     testing.assert_array_equal(binary_res, grey_res)
 
+def test_binary_perimeter():
+    bw = np.zeros([10,10])
+    bw[3:7,3:7] = 1
+    perim = np.copy(bw)
+    perim[4:6,4:6] = 0
+    strel = selem.square(3)
+    binary_res = binary.binary_perimeter(bw, strel)
```
Comment (emmanuelle) [inline skimage/morphology/tests/test_binary.py]: you could also write the test with the default structuring element, the result would be the same.

File: skimage/morphology/binary.py (hunk 3 above)
Comment (emmanuelle) [inline skimage/morphology/binary.py]: +1. In the Notes section, please explain the algorithm: the perimeter is the original shape minus its erosion.

File: skimage/morphology/binary.py [hunk 6]
```diff
@@ -144,3 +144,35 @@ def binary_closing(image, selem=None, out=None):
...
+        Binary input image.
+    selem : ndarray, optional
+        The neighborhood expressed as a N-D array of 1's and 0's
+        where N is the dimensionality of image.
+        If None, use cross-shaped structuring element (connectivity=1).
+
+    Returns
+    -------
+    perim : binary image of perimeters of objects in bw
+
```
Comment (emmanuelle) [inline skimage/morphology/binary.py]: Please add an Examples section.

File: skimage/morphology/binary.py [hunk 7]
```diff
@@ -144,3 +144,35 @@ def binary_closing(image, selem=None, out=None):
...
+    """
+
+    bw_pad = np.pad(image, pad_width=1, mode='constant', constant_values=0)
```
Comment (emmanuelle) [inline skimage/morphology/binary.py]: Are you sure that padding is necessary? ``binary_erosion`` erodes border pixels as well.

**Comment by jacobkimmel** [comment]
@jnl it seems `segmentation.find_boundaries` will achieve the same result as the example I listed above, though it does so by a slightly different algorithm.
Likely, this PR would be redundant. I'll close it.

As a note -- I searched for a decent amount of time before submitting this PR without managing to identify `find_boundaries` as a relevant function.
Perhaps I am alone in this, but `segmentation` is not an obvious place to look for a morphological transformation of a binary image, especially given that both `binary` and `morphology` exist!

It may be helpful if we add an example to the docs running through some of the standard binary operations users migrating from other toolboxes may be familiar with. 
I'm not sure of skimage's policy toward didactic docs, but a basic overview of binary operations may be useful for users new to image processing as well. 
If this sounds useful, let me know and I can draft something. If not, no offense taken!

Thanks for the feedback everyone, cheers!

**Comment by jni** [comment]
@jacobkimmel Sorry about that! Discoverability is something that we need to work on, for sure. I've advocated for a flat namespace before, while others have suggested cross-importing functions (so there would be both `morphology.find_boundaries` and `segmentation.find_boundaries`). The tip that killed the discussion at the time was a little-known NumPy trick (which we should probably add to the [user guide](http://scikit-image.org/docs/dev/user_guide.html) somewhere!):

```python
import numpy as np
import skimage
np.lookfor('boundaries', skimage)
```

At any rate, documentation improvements are very much appreciated! To add an example to the [gallery](http://scikit-image.org/docs/dev/auto_examples/), all you need to do is add a file in [`doc/examples/*/plot_*.py`](https://github.com/scikit-image/scikit-image/tree/master/doc/examples). Look at other files in there as a formatting template.

Finally, please don't hesitate to subscribe and write to the [mailing list](http://scikit-image.org/docs/dev/user_guide/getting_help.html#mailing-list), which should be your first port of call for getting help using skimage.

Thanks for your time and sorry again!

**Pull request closed without merging**